### PR TITLE
[Serve] Refactor Metric System: Counter + Measure Support

### PR DESCRIPTION
**Pull request by simon-mo**
This PR reimplements a metric system for RayServe. Each actors will push metrics to the sink actor. The sink actor can be `InMemorySink` or `PrometheusSink`. Both sinks are exposed through `serve.stat` as well as `/-/metrics` API.

The interface for adding metrics is:
```
client = MetricClient.connect_from_serve()
counter = client.new_counter(...)
counter.add()
```

Our metrics support arbitrary labels to be applied ahead of time as well as at runtime.

Won't do in this PR:
- More aggregation of measures. Currently we only keep the latest measure
- More instrumentation of each Serve components. Currently we only have few counters to demonstrate the metric capability. 
- Checkpoint and recovery of the sink. The sink is reconstructable but it will loss states in between.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

**Comment by edoakes** [comment]
@simon-mo Can we make this component fault tolerant in this PR? Basically just need to set its max_reconstructions and then make sure all initialization is done in the constructor. If it needs any state, it should be fetched from the master actor on startup.

File: python/ray/serve/api.py [hunk 1]
```diff
@@ -141,10 +140,11 @@ def kv_store_connector(namespace):
     master_actor = ServeMaster.options(
         detached=True, name=SERVE_MASTER_NAME).remote(kv_store_connector)
 
+    ray.get(master_actor.start_metric_sink.remote())
```
Comment (edoakes) [inline python/ray/serve/api.py:143]: Let's move this to the constructor

**Comment by edoakes** [comment]
@simon-mo could you update the description to specify what this *does* do?

File: python/setup.py [hunk 2]
```diff
@@ -77,7 +77,7 @@
 extras = {
     "debug": [],
     "dashboard": ["requests"],
-    "serve": ["uvicorn", "pygments", "werkzeug", "flask", "pandas", "blist"],
+    "serve": ["uvicorn", "flask", "blist", "prometheus_client"],
```
Comment (edoakes) [inline python/setup.py:80]: This should probably be an optional dependency only for those who are exporting to prometheus

File: python/ray/serve/api.py [hunk 3]
```diff
@@ -314,16 +312,7 @@ def get_handle(endpoint_name,
...
-        percentiles(List[int]): The percentiles for aggregation operations.
-            Default is 50th, 90th, 95th percentile.
-        agg_windows_seconds(List[int]): The aggregation windows in seconds.
-            The longest aggregation window must be shorter or equal to the
-            gc_window_seconds.
-    """
-    [monitor] = ray.get(master_actor.get_metric_monitor.remote())
-    return ray.get(monitor.collect.remote(percentiles, agg_windows_seconds))
+def stat():
+    """Retrieve metric statistics about ray serve system."""
```
Comment (edoakes) [inline python/ray/serve/api.py:316]: Should specify what's returned in the dictionary here

File: python/ray/serve/backend_worker.py [hunk 4]
```diff
@@ -82,24 +79,11 @@ def __init__(self, name, _callable, router_handle, is_function):
...
-            },
-            "{}_latency_s".format(self.name): {
-                "value": latency_list,
-                "type": "list",
-            },
-        }
+        self.metric_collector = PushCollector.connect_from_serve()
+        self.error_counter = self.metric_collector.new_counter(
+            "backend_error_counter",
+            description="Number of exceptions occurred in backend",
```
Comment (edoakes) [inline python/ray/serve/backend_worker.py:85]:
```suggestion
            description="Number of exceptions that have occurred in the backend",
```

File: python/ray/serve/backend_worker.py [hunk 5]
```diff
@@ -82,24 +79,11 @@ def __init__(self, name, _callable, router_handle, is_function):
...
+            labels={"backend": self.name})
```
Comment (edoakes) [inline python/ray/serve/backend_worker.py:86]: I think self.name is a replica tag, right? We should probably change the key here to `replica` or `replica_tag`

File: python/ray/serve/backend_worker.py [hunk 6]
```diff
@@ -82,24 +79,11 @@ def __init__(self, name, _callable, router_handle, is_function):
...
-                "value": self.error_counter,
-                "type": "counter",
-            },
-            "{}_latency_s".format(self.name): {
-                "value": latency_list,
-                "type": "list",
-            },
-        }
+        self.metric_collector = PushCollector.connect_from_serve()
+        self.error_counter = self.metric_collector.new_counter(
```
Comment (edoakes) [inline python/ray/serve/backend_worker.py:83]: Don't we also want to count the number of requests successfully served?

File: python/ray/serve/backend_worker.py [hunk 7]
```diff
@@ -216,7 +196,7 @@ async def invoke_batch(self, request_item_list):
             return result_list
         except Exception as e:
             wrapped_exception = wrap_to_ray_error(e)
-            self.error_counter += batch_size
+            self.error_counter.add(batch_size)
```
Comment (edoakes) [inline python/ray/serve/backend_worker.py:199]: This doesn't match the description of the counter - "number of exceptions that have occurred in the backend." Should either only add one here or update the description to specify that it's the number of inputs that this has errored on.

File: python/ray/serve/constants.py [hunk 8]
```diff
@@ -18,3 +18,6 @@
 
 #: Key for storing no http route services
 NO_ROUTE_KEY = "NO_ROUTE"
+
+#: Interval in sections for pushing metrics to the metric sink
```
Comment (edoakes) [inline python/ray/serve/constants.py:22]:
```suggestion
#: Interval in seconds for pushing metrics to the metric sink
```

File: python/ray/serve/http_proxy.py [hunk 9]
```diff
@@ -26,11 +26,21 @@ class HTTPProxy:
...
+        master = _get_master_actor()
+        self.route_table, [self.router_handle
+                           ] = await master.get_http_proxy_config.remote()
+        [self.metric_sink] = await master.get_metric_sink.remote()
+
+        self.metric_collector = PushCollector.connect_from_serve()
+        self.request_counter = self.metric_collector.new_counter(
+            "num_http_requests",
+            description="The number of requests processed",
+        )
```
Comment (edoakes) [inline python/ray/serve/http_proxy.py:43]: These should probably go in `__init__` (unless this needs to be in an async context, in which case we should rename this method)

File: python/ray/serve/master.py [hunk 10]
```diff
@@ -109,17 +111,16 @@ def get_http_proxy(self):
...
 
-    def start_metric_monitor(self, gc_window_seconds):
-        assert self.metric_monitor is None, "Metric monitor already started."
-        self.metric_monitor = MetricMonitor.remote(gc_window_seconds)
-        # TODO(edoakes): this should be an actor method, not a separate task.
-        start_metric_monitor_loop.remote(self.metric_monitor)
-        self.metric_monitor.add_target.remote(self.router)
+    def start_metric_sink(self):
+        assert self.metric_sink is None, "Metric sink already started."
+        self.metric_sink = async_retryable(PrometheusSinkActor).options(
```
Comment (edoakes) [inline python/ray/serve/master.py:116]: Should this be `num_cpus=0` like the old one?

File: python/ray/serve/metric.py [hunk 11]
```diff
@@ -1,157 +1,178 @@
...
+    def record(self, value):
+        self.collector.push_event(EventType.MEASURE, self.name, value)
+
+
+class EventType(enum.IntEnum):
+    COUNTER = 1
+    MEASURE = 2
+
+    def to_user_facing_class(self):
+        return {self.COUNTER: Counter, self.MEASURE: Measure}[self.value]
```
Comment (edoakes) [inline python/ray/serve/metric.py:95]: This will construct a new dictionary every time this is called and then index into it, right? That seems pretty slow - why not just an `if/else`?

File: python/ray/serve/metric.py [hunk 12]
```diff
@@ -1,157 +1,178 @@
...
+        return {self.COUNTER: PromCounter, self.MEASURE: PromGauge}[self.value]
+
+
+MetricMetadata = Dict[str, Dict]
+MetricEventBatch = List[Tuple[EventType, str, float]]
+
+
+class PushCollector:
+    def __init__(self,
+                 push_batch_callback: Callable[
```
Comment (edoakes) [inline python/ray/serve/metric.py:107]: Is the `push_batch_callback` only used for testing right now? Fine if it is, but I think it would probably be simpler to pass in an object that inherits from some `MetricClient`

File: python/ray/serve/metric.py [hunk 13]
```diff
@@ -1,157 +1,178 @@
...
+
+    async def push_once(self):
+        old_batch, self.metric_events = self.metric_events, []
+        self.new_metric_added.clear()
+        await self.push_batch_callback(self.metric_metadata, old_batch)
+
+    async def push_forever(self, interval_s):
+        while True:
+            await self.push_once()
+            await asyncio.wait(
```
Comment (edoakes) [inline python/ray/serve/metric.py:175]: I'm a bit confused by this logic - so we either sleep for `interval` seconds or wake up anytime there's a new metric added? Won't we not have any batching in that case because it pushes every time a metric is added? Also, what's the point of the sleep at all if we wake up anytime a new metric is added?

**Review by edoakes — COMMENTED** [review]
Do we want the prometheus exporter to be on by default? To me it seems like it would make more sense to have a simple, dependency-less version where the sink is just an actor that returns a dictionary of metrics when pinged.

**Comment by rkooo567** [comment]
@simon-mo Is it going to be included in the next release?

**Comment by simon-mo** [comment]
Not release blocking.

File: python/ray/serve/api.py [hunk 14]
```diff
@@ -96,12 +99,14 @@ def init(
...
-            memory. Data older than the gc_window will be deleted. The default
-            is 3600 seconds, which is 1 hour.
         queueing_policy(RoutePolicy): Define the queueing policy for selecting
             the backend for a service. (Default: RoutePolicy.Random)
         policy_kwargs: Arguments required to instantiate a queueing policy
+        metric_sink(BaseSink): The metric storage actor for all RayServe actors
+            to push to. RayServe have two options built in: InMemorySink and
+            PrometheusSink
+        metric_push_interval(float): The interval for each actors to push to
+            the metric sink. Default is 5s.
```
Comment (edoakes) [inline python/ray/serve/api.py:109]:
```suggestion
            the metric sink. Default is 5s.
```
```suggestion
            the metric sink. Default is 2s.
```

Comment (edoakes) [reply python/ray/serve/api.py:109]: It's set to 2s below in `constants.py`. Is there any reason for having a constant and not just hardcoding it here?

File: python/ray/serve/http_proxy.py [hunk 15]
```diff
@@ -179,7 +204,8 @@ async def run(self):
         sock.bind((self.host, self.port))
         sock.set_inheritable(True)
 
-        config = uvicorn.Config(self.app, lifespan="on", access_log=False)
+        config = uvicorn.Config(
+            self.app, lifespan="on", access_log=False, debug=True)
```
Comment (edoakes) [inline python/ray/serve/http_proxy.py:208]: What's the reason for `debug=True`? Should this always be on?

File: python/ray/serve/master.py [hunk 16]
```diff
@@ -69,6 +72,8 @@ async def __init__(self, kv_store_connector, router_class, router_kwargs,
         # Dictionary of backend tag to dictionaries of replica tag to worker.
         # TODO(edoakes): consider removing this and just using the names.
         self.workers = defaultdict(dict)
+        # Dictionary to keep track of mutable system wide config
```
Comment (edoakes) [inline python/ray/serve/master.py:75]:
```suggestion
        # Dictionary to keep track of mutable, system-wide config.
```

Comment (edoakes) [reply python/ray/serve/master.py:75]: This is not currently mutable, right?

**Review by edoakes — COMMENTED** [review]
I think this looks reasonable, but how will we support push-based monitoring? The changes here assume that the user wants a metric sink running in their serve cluster, but for the push-based case they probably just want a hook to provide their own client. I think we should nail down that API before merging this.

File: python/ray/serve/api.py [hunk 17]
```diff
@@ -96,12 +99,14 @@ def init(
...
             connection. Default to {"object_store_memory": int(1e8)} for
             performance stability reason
-        gc_window_seconds(int): How long will we keep the metric data in
-            memory. Data older than the gc_window will be deleted. The default
-            is 3600 seconds, which is 1 hour.
         queueing_policy(RoutePolicy): Define the queueing policy for selecting
             the backend for a service. (Default: RoutePolicy.Random)
         policy_kwargs: Arguments required to instantiate a queueing policy
+        metric_sink(BaseSink): The metric storage actor for all RayServe actors
+            to push to. RayServe have two options built in: InMemorySink and
```
Comment (edoakes) [inline python/ray/serve/api.py:106]:
```suggestion
            to push to. RayServe has two options built in: InMemorySink and
```

File: python/ray/serve/metric/client.py [hunk 18]
```diff
@@ -0,0 +1,156 @@
...
+
+
+class MetricClient:
+    def __init__(
+            self,
+            metric_sink_actor,
+            push_interval: float,
+            default_labels: Optional[Dict[str, str]] = None,
+    ):
+        """Initialize a client to push metric to sink.
```
Comment (edoakes) [inline python/ray/serve/metric/client.py:23]:
```suggestion
        """Initialize a client to push metrics to the sink actor.
```

File: python/ray/serve/metric/client.py [hunk 19]
```diff
@@ -0,0 +1,156 @@
...
+
+        Args:
+            metric_sink_actor: The actor to push metric batch to.
```
Comment (edoakes) [inline python/ray/serve/metric/client.py:26]:
```suggestion
            metric_sink_actor: The actor to push metrics to.
```

File: python/ray/serve/metric/client.py [hunk 20]
```diff
@@ -0,0 +1,156 @@
...
+            default_labels(dict): The set of labels to apply for all metrics
+                created by this actor. For example, {"source": "worker"}
```
Comment (edoakes) [inline python/ray/serve/metric/client.py:28]:
```suggestion
                created by this actor. For example, {"source": "worker"}.
```

File: python/ray/serve/metric/client.py [hunk 21]
```diff
@@ -0,0 +1,156 @@
...
+        self.metric_records = []
+
+        assert asyncio.get_event_loop().is_running()
+        self.push_task = asyncio.get_event_loop().create_task(
+            self.push_forever(push_interval))
+        logger.debug("Initialized client")
+
+    @staticmethod
+    def connect_from_serve(default_labels: Optional[Dict[str, str]] = None):
+        """Create the metric client automatically when running inside serve"""
```
Comment (edoakes) [inline python/ray/serve/metric/client.py:43]:
```suggestion
        """Create the metric client automatically when running inside serve."""
```

File: python/ray/serve/metric/client.py [hunk 22]
```diff
@@ -0,0 +1,156 @@
...
+                description="This is a simple measure for latency in ms",
+                label_names=("route"))
+        >>> measure.labels(route="/hi").record(42)
+        """
+        return self._new_metric(name, MetricType.MEASURE, description,
+                                label_names)
+
+    def record_event(self, record: MetricRecord):
+        """Called by Counter or Measure to record one observation"""
+        self.metric_records.append(record)
```
Comment (edoakes) [inline python/ray/serve/metric/client.py:113]: This doesn't seem like it should be a function

File: python/ray/serve/metric/client.py [hunk 23]
```diff
@@ -0,0 +1,156 @@
...
+        )
+        metric_class = convert_event_type_to_class(metric_type)
+        metric_object = metric_class(
+            client=self, name=name, label_names=label_names)
+
+        self.registered_metrics[name] = metric_metadata
+        return metric_object
+
+    async def _push_once(self):
+        if not len(self.metric_records):
```
Comment (edoakes) [inline python/ray/serve/metric/client.py:145]: nit: I prefer len() == 0 for clarity given that this isn't really any more concise

File: python/ray/serve/metric/client.py [hunk 24]
```diff
@@ -0,0 +1,156 @@
...
+            return
+
+        old_batch, self.metric_records = self.metric_records, []
+        logger.debug("Pushing metric batch {}".format(old_batch))
+        await retry_actor_failures_async(self.sink.push_batch,
+                                         self.registered_metrics, old_batch)
+
+    async def push_forever(self, interval_s):
+        while True:
+            await self._push_once()
```
Comment (edoakes) [inline python/ray/serve/metric/client.py:155]:
```suggestion
            await self._push_batch()
```
?

File: python/ray/serve/metric/sink.py [hunk 25]
```diff
@@ -0,0 +1,139 @@
...
+    fields += list(record.labels.keys())
+
+    merged_labels = metadata.default_labels.copy()
+    merged_labels.update(record.labels)
+
+    tuple_type = namedtuple(metadata.name, fields)
+    return tuple_type(name=metadata.name, type=metadata.type, **merged_labels)
+
+
+class BaseSink:
```
Comment (edoakes) [inline python/ray/serve/metric/sink.py:23]: I don't really like the name `BaseSink` because it doesn't communicate that this is actually to be used as a client (either pull- or push-based). How about something like `MetricsExporter`?

File: python/ray/serve/metric/sink.py [hunk 26]
```diff
@@ -0,0 +1,139 @@
...
+    """A sink is the place to store or forward metrics to external services"""
```
Comment (edoakes) [inline python/ray/serve/metric/sink.py:24]:
```suggestion
    """Metrics are aggregated pushed from other actors in the system to this actor. They can then be stored or pushed to an external monitoring system."""
```

File: python/ray/serve/metric/sink.py [hunk 27]
```diff
@@ -0,0 +1,139 @@
...
+
+    def __init__(self):
+        self.metadata = dict()
+        self.recording = []
```
Comment (edoakes) [inline python/ray/serve/metric/sink.py:28]: This name doesn't really tell me what it holds out of context. Maybe `self.metric_batches`?

File: python/ray/serve/metric/sink.py [hunk 28]
```diff
@@ -0,0 +1,139 @@
...
+
+    tuple_type = namedtuple(metadata.name, fields)
+    return tuple_type(name=metadata.name, type=metadata.type, **merged_labels)
+
+
+class BaseSink:
+    """A sink is the place to store or forward metrics to external services"""
+
+    def __init__(self):
+        self.metadata = dict()
```
Comment (edoakes) [inline python/ray/serve/metric/sink.py:27]: What is the metadata used for? Can we recover this after failure?

File: python/ray/serve/metric/sink.py [hunk 29]
```diff
@@ -0,0 +1,139 @@
...
+        self.recording = []
+        logger.debug("Sink initialized {}".format(type(self)))
+
+    def push_batch(self, metadata: Dict[str, MetricMetadata],
+                   batch: MetricBatch):
+        self.metadata.update(metadata)
+        self.recording.extend(batch)
+        self.compact()
+
+    def compact(self):
```
Comment (edoakes) [inline python/ray/serve/metric/sink.py:37]: `compact` sounds like a specific implementation of of this method rather than a description of what's to be done. How about `process_metric_batches`?

File: python/ray/serve/metric/sink.py [hunk 30]
```diff
@@ -0,0 +1,139 @@
...
+        raise NotImplementedError(
+            "This method should be implemented by subclass.")
+
+    def get_metric(self):
```
Comment (edoakes) [inline python/ray/serve/metric/sink.py:41]:
```suggestion
    def get_metrics(self):
```

Comment (edoakes) [reply python/ray/serve/metric/sink.py:41]: It seems sort of strange to me that both this and `compact` would be implemented by the same class. One is storing metrics and serving them, one is just pushing them out. Doesn't seem like the concepts should be combined in one interface.

File: python/ray/serve/metric/sink.py [hunk 31]
```diff
@@ -0,0 +1,139 @@
...
+    def __init__(self):
+        super().__init__()
+
+        # Keep track of counters
+        self.counters: Counter[namedtuple, float] = Counter()
+        # Keep track of latest observation of measures
+        self.latest_measures: Dict[namedtuple, float] = dict()
+
+    def compact(self):
+        for record in self.recording:
```
Comment (edoakes) [inline python/ray/serve/metric/sink.py:57]: I'm not a huge fan of this inheritance-based pattern because it's hard to follow what exactly `compact` should do and what is done by the superclass.

How about changing this to users providing a class that implements `compact` that takes the latest recordings as input, then always clear them after calling `compact` in the caller.

File: python/ray/serve/metric/sink.py [hunk 32]
```diff
@@ -0,0 +1,139 @@
...
+        from prometheus_client import generate_latest
+        return generate_latest(self.registry)
+
+    def compact(self):
+        self._process_metadata(self.metadata)
+        self._process_batch(self.recording)
+        self.recording = []
+
+    def _metric_type_to_prometheus_class(self, metric_type: MetricType):
+        from prometheus_client import Counter, Gauge
```
Comment (edoakes) [inline python/ray/serve/metric/sink.py:99]: Can we put all of the prometheus_client lazy imports in the constructor?

File: python/ray/serve/api.py [hunk 33]
```diff
@@ -57,19 +61,18 @@ def __call__(self, *, python_arg=None):
...
+         start_server=True,
+         http_host=DEFAULT_HTTP_HOST,
+         http_port=DEFAULT_HTTP_PORT,
+         ray_init_kwargs={
+             "object_store_memory": int(1e8),
+             "num_cpus": max(cpu_count(), 8)
+         },
+         queueing_policy=RoutePolicy.Random,
+         policy_kwargs={},
+         metric_sink=InMemorySink,
```
[the remark this answers is not in the document]
Comment (simon-mo) [reply python/ray/serve/api.py:74]: there's a separate problem to solve for this, namely making sure we can make all configs string based and also supporting arbitrary Python functions/class to queuing policy or metric exporter implementation. 

out of scope for this PR, but I will for sure add a backlog

File: python/ray/serve/metric/sink.py (hunk 31 above)
Comment (simon-mo) [reply python/ray/serve/metric/sink.py:57]: changed the signature of compact to `export(self, metric_metadata, metric_batch)`, this should give a clear idea that the entire batch should be consumed.

File: python/ray/serve/metric/sink.py (hunk 30 above)
Comment (simon-mo) [reply python/ray/serve/metric/sink.py]: changed this to `inspect_metrics` which makes more sense since this method will only be called by `serve.stat` or `/-/metrics`

File: python/ray/serve/api.py [hunk 34]
```diff
@@ -57,19 +61,17 @@ def __call__(self, *, python_arg=None):
...
+         start_server=True,
+         http_host=DEFAULT_HTTP_HOST,
+         http_port=DEFAULT_HTTP_PORT,
+         ray_init_kwargs={
+             "object_store_memory": int(1e8),
+             "num_cpus": max(cpu_count(), 8)
+         },
+         queueing_policy=RoutePolicy.Random,
+         policy_kwargs={},
+         metric_exporter=InMemoryExporter):
```
Comment (edoakes) [inline python/ray/serve/api.py:74]: We will need to make this configurable during runtime in the future, but this seems fine for now.

File: python/ray/serve/backend_worker.py [hunk 35]
```diff
@@ -71,32 +70,26 @@ def __init__(self, name, _callable, is_function):
...
-            "{}_error_counter".format(self.name): {
-                "value": self.error_counter,
-                "type": "counter",
-            },
-            "{}_latency_s".format(self.name): {
-                "value": latency_list,
-                "type": "list",
-            },
-        }
+        self.metric_client = MetricClient.connect_from_serve(
```
Comment (edoakes) [inline python/ray/serve/backend_worker.py:73]: Can we also add a counter for the number of times each replica has been started? This would help monitor if workers are crashing.

File: python/ray/serve/metric/client.py [hunk 36]
```diff
@@ -0,0 +1,151 @@
...
+            self.push_to_exporter_forever(push_interval))
+        logger.debug("Initialized client")
+
+    @staticmethod
+    def connect_from_serve(default_labels: Optional[Dict[str, str]] = None):
+        """Create the metric client automatically when running inside serve."""
+        from ray.serve.api import _get_master_actor
+
+        master_actor = _get_master_actor()
+        [metric_exporter, push_interval] = retry_actor_failures(
```
Comment (edoakes) [inline python/ray/serve/metric/client.py:46]: The push_interval is just a constant now, right? Let's just move the constant here

File: python/ray/serve/master.py [hunk 37]
```diff
@@ -215,8 +213,8 @@ async def _recover_from_checkpoint(self, checkpoint_bytes):
         # Load internal state from the checkpoint data.
         (self.routes, self.backends, self.traffic_policies, self.replicas,
          self.replicas_to_start, self.replicas_to_stop,
-         self.backends_to_remove,
-         self.endpoints_to_remove) = pickle.loads(checkpoint_bytes)
+         self.backends_to_remove, self.endpoints_to_remove,
+         self.config) = pickle.loads(checkpoint_bytes)
```
[the remark this answers is not in the document]
Comment (edoakes) [reply python/ray/serve/master.py:216]: I think this is the only instance of `self.config` now - did you forget to remove this or am I missing something?

File: python/ray/serve/metric/exporter.py [hunk 38]
```diff
@@ -0,0 +1,144 @@
...
+    """Aggregate metrics from all RayServe actors and export them.
+
+    Metrics are aggregated pushed from other actors in the system to
+    this actor. They can then be stored or pushed to an external monitoring
+    system.
+    """
+
+    def __init__(self):
+        # Stores the mapping metric_name -> MetricMetadata
+        # This field is tolerant to failures since each client will always push
```
Comment (edoakes) [inline python/ray/serve/metric/exporter.py:35]: Will each client push an updated copy of the metadata with every message?

File: python/ray/serve/metric/exporter.py [hunk 39]
```diff
@@ -0,0 +1,144 @@
...
+        raise NotImplementedError(
+            "This method should be implemented by subclass.")
+
+    def inspect_metrics(self):
+        raise NotImplementedError(
+            "This method should be implemented by subclass.")
+
+
+@ray.remote(num_cpus=0)
+class InMemoryExporter(BaseExporter):
```
Comment (edoakes) [inline python/ray/serve/metric/exporter.py:57]: Can we use composition instead of inheritance here? I don't think inheritance buys us anything and is a worse user interface IMO

File: python/ray/serve/metric/exporter.py [hunk 40]
```diff
@@ -0,0 +1,144 @@
...
+                self.latest_measures[metric_key] = record.value
+            else:
+                raise RuntimeError("Unrecognized metric type {}".format(
+                    metadata.type))
+
+    def inspect_metrics(self):
+        items = []
+        metrics_to_collect = {**self.counters, **self.latest_measures}
+        for info_tuple, value in metrics_to_collect.items():
+            # Represent the metric type in human readable name
```
Comment (edoakes) [inline python/ray/serve/metric/exporter.py:82]:
```suggestion
            # Represent the metric type as a human readable name.
```

File: python/ray/serve/router.py [hunk 41]
```diff
@@ -166,22 +168,20 @@ async def __init__(self):
         for backend, backend_config in backend_configs.items():
             await self.set_backend_config(backend, backend_config)
 
+        self.metric_client = MetricClient.connect_from_serve()
+        self.num_router_requests = self.metric_client.new_counter(
+            "num_router_requests",
+            description="Number of requests from the router's view.",
```
Comment (edoakes) [inline python/ray/serve/router.py:174]:
```suggestion
            description="Number of requests processed by the router.",
```

File: python/ray/serve/router.py [hunk 42]
```diff
@@ -202,7 +202,10 @@ async def enqueue_request(self, request_meta, *request_args,
 
         # Note: a future change can be to directly return the ObjectID from
         # replica task submission
-        result = await query.async_future
+        try:
+            result = await query.async_future
+        except RayTaskError as e:
+            result = e
```
Comment (edoakes) [inline python/ray/serve/router.py:208]: Should we count these too? Can be aggregated from others of course but might be nice to have a first-class global counter.

File: python/ray/serve/tests/conftest.py [hunk 43]
```diff
@@ -16,10 +15,5 @@ def serve_instance():
...
 @pytest.fixture(scope="session")
-def ray_instance():
-    ray_already_initialized = ray.is_initialized()
-    if not ray_already_initialized:
-        ray.init(object_store_memory=int(1e8))
-    yield
-    if not ray_already_initialized:
-        ray.shutdown()
+def ray_instance(serve_instance):
+    pass
```
Comment (edoakes) [inline python/ray/serve/tests/conftest.py:19]: Why this change? Shouldn't we just use serve_instance in this case?

**Comment by edoakes** [comment]
@simon-mo the naming looks much better now! Just some last cleanup comments before lgtm'ing. Looks like `test_serve` is failing in CI as well.

File: python/ray/serve/metric/exporter.py (hunk 38 above)
Comment (simon-mo) [reply python/ray/serve/metric/exporter.py:35]: Yes

File: python/ray/serve/tests/conftest.py (hunk 43 above)
Comment (simon-mo) [reply python/ray/serve/tests/conftest.py:19]: Yeah the change was to avoid the segfault warnings, i just replaced all ray_instance with serve_instance

**Review by edoakes — APPROVED** [review]
LGTM!
